massa-deferred-calls/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+use massa_db_exports::{
+    DBBatch, ShareableMassaDBController, CRUD_ERROR, DEFERRED_CALLS_PREFIX,
+    DEFERRED_CALL_DESER_ERROR, DEFERRED_CALL_SER_ERROR, KEY_DESER_ERROR, STATE_CF,
+};
+use massa_serialization::{DeserializeError, Deserializer, Serializer};
+use registry_changes::{
+    DeferredRegistryChanges, DeferredRegistryChangesDeserializer, DeferredRegistryChangesSerializer,
+};
+
+/// This module implements a new version of the Autonomous Smart Contracts. (ASC)
+/// This new version allow asynchronous calls to be registered for a specific slot and ensure his execution.
+mod call;
+pub mod registry_changes;
+pub mod slot_changes;
+
+#[cfg(test)]
+mod tests;
+
+#[macro_use]
+mod macros;
+
+pub use call::DeferredCall;
+use massa_ledger_exports::{SetOrDelete, SetOrKeep};
+use massa_models::{
+    amount::Amount,
+    config::{DEFERRED_CALL_MAX_POOL_CHANGES, MAX_ASYNC_GAS, THREAD_COUNT},
+    deferred_call_id::{DeferredCallId, DeferredCallIdDeserializer, DeferredCallIdSerializer},
+    slot::Slot,
+};
+use std::collections::BTreeMap;
+
+// #[derive(Debug)]
+pub struct DeferredCallRegistry {
+    db: ShareableMassaDBController,
+    call_serializer: DeferredCallSerializer,
+    call_id_serializer: DeferredCallIdSerializer,
+    call_deserializer: DeferredCallDeserializer,
+    call_id_deserializer: DeferredCallIdDeserializer,
+    registry_changes_deserializer: DeferredRegistryChangesDeserializer,
+    registry_changes_serializer: DeferredRegistryChangesSerializer,
+}
+
+impl DeferredCallRegistry {
+    /*
+     DB layout:
+        [DEFERRED_CALL_TOTAL_GAS] -> u64 // total currently booked gas
+        [DEFERRED_CALLS_PREFIX][slot][SLOT_TOTAL_GAS] -> u64 // total gas booked for a slot (optional, default 0, deleted when set to 0)
+        [DEFERRED_CALLS_PREFIX][slot][SLOT_BASE_FEE] -> u64 // deleted when set to 0
+        [DEFERRED_CALLS_PREFIX][slot][CALLS_TAG][id][CALL_FIELD_X_TAG] -> AsyncCall.x // call data
+    */
+
+    // TODO pass args
+    pub fn new(db: ShareableMassaDBController) -> Self {
+        Self {
+            db,
+            call_serializer: DeferredCallSerializer::new(),
+            call_id_serializer: DeferredCallIdSerializer::new(),
+            call_deserializer: DeferredCallDeserializer::new(THREAD_COUNT),
+            call_id_deserializer: DeferredCallIdDeserializer::new(),
+            registry_changes_deserializer: DeferredRegistryChangesDeserializer::new(
+                THREAD_COUNT,
+                MAX_ASYNC_GAS,
+                DEFERRED_CALL_MAX_POOL_CHANGES,
+            ),
+            registry_changes_serializer: DeferredRegistryChangesSerializer::new(),
+        }
+    }
+
+    pub fn get_slot_calls(&self, slot: Slot) -> DeferredSlotCalls {
+        let mut to_return = DeferredSlotCalls::new(slot);
+        let key = deferred_slot_call_prefix_key!(slot.to_bytes_key());
+
+        // cache the call ids to avoid duplicates iteration
+        let mut temp = Vec::new();
```

Comment:
More general question: I don't think this check is necessary: there should never be duplicates in the first place